cmd/talosctl/cmd/talos/root.go

```diff
@@ -123,3 +127,92 @@ var Commands []*cobra.Command
 func addCommand(cmd *cobra.Command) {
 	Commands = append(Commands, cmd)
 }
+
+// completeResource represents tab complete options for `ls` and `ls *` commands.
+func completePathFromNode(inputPath string) []string {
+	pathToSearch := inputPath
+
+	// If the pathToSearch is empty, use root '/'
+	if pathToSearch == "" {
+		pathToSearch = "/"
+	}
+
+	var paths map[string]struct{}
+
+	// search up one level to find possible completions
+	if pathToSearch != "/" && !strings.HasSuffix(pathToSearch, "/") {
+		splits := strings.Split(pathToSearch, "/")
+		// we need a trailing slash to search for items in a directory
+		pathToSearch = strings.Join(splits[:len(splits)-1], "/") + "/"
+	}
+
+	paths = getPathFromNode(pathToSearch)
+
+	result := make([]string, 0, len(paths))
+
+	for k := range paths {
+		result = append(result, k)
+	}
+
+	return filterPaths(result, inputPath)
+}
+
+func filterPaths(input []string, filter string) []string {
+	result := make([]string, 0, len(input))
+
+	for _, v := range input {
+		if strings.HasPrefix(v, filter) {
+			result = append(result, v)
+		}
+	}
+
+	return result
+}
+
+//nolint:gocyclo
+func getPathFromNode(path string) map[string]struct{} {
+	paths := make(map[string]struct{})
+
+	if WithClient(func(ctx context.Context, c *client.Client) error {
+		stream, err := c.LS(ctx, &machineapi.ListRequest{
```

Comment:
related to the comment below, we could also make sure we abort the `LS` call on the server if we exit early:

```suggestion
                ctx, cancel := context.WithCancel(ctx)
                defer cancel()

		stream, err := c.LS(ctx, &machineapi.ListRequest{
```

this way if we exit this function, we always abort the API on the server side (so it stops listing)

Reply:
fixed